HW0/P5/P5.txt

```diff
@@ -1 +1,146 @@
+time). 
+
+*********************************
+In our case, it takes 22 seconds to total 256 bags. 
+*********************************
+
+Part 6. (Answer)
+
+(One Bag Case) Assume, initially, that each employee can only grab one bag, that this transaction 
+takes one second and that I can only interact with one person at a time. Then it will take 256
+seconds to distribute the bags from me to 256 employees.
+
+*********************************
+In this case, distributing the bags to be added (using whatever method) is slower than adding
+them with one person. 
+*********************************
+
+(Two Bags Case) If each employee can grab two bags, then it takes 128 seconds to distribute the 
+bags. Assuming adding doesn't commence until the bags are distributed, thereafter it will take
+3*ceil(log_2(N)) - 2 seconds to total the bags. So, together it will take 
+ceil(N/2) + 3*ceil(log_2(N)) - 2 
+number of seconds to total N number of bags. This is still faster than computing the total with 
+one employee. 
+
+Suppose adding starts when the first two bags are distributed (and that, in case N is odd, the last
+employee receives just one bag). Then the pairs of summands for the first set of subtotals will be 
+available at time t=1, t=2, ..., t=ceil(N/2) respectively (i.e. the nth pair of bags is distributed, 
+and become available for summing, at t=n). We'd like to calculate a set of subtotal from our pairs 
+of bags as soon as possible. We can verify that this means having each employee sum up the pair of 
+bags each receives. So, the soonest by which the first set of subtotals can be available is t=2, 
+t=3, ..., t=ceil(N/2)+1 (the subtotal for the nth pair of bags we hand out will be computed by time 
+t=n+1) respectively.  
+
+Now, suppose the k-th set of subtotals are optimally computed. The fastest way to compute (k+1)-th 
+set of subtotals is to sum up pairs of summands with the earliest available time. That is, we sort
+the list of k-th subtotals by availability time, and pairs of consecutive items in the list become
+the summands for the (k+1)-th subtotals. So, each (k+1)-th subtotal is computed by time 
+t = max{t_s1, t_s2} + 2, where t_s1, t_s2 are the times at which summand 1 and summand 2 become 
+available, respectively. The rationale is: assuming t_s2 > t_s1, and that it takes one second to
+receive each summand, summand 1 will already have been received when summand 2 become available.
+Then, it will take one additional second to receive summand 2 and one more second to add them.
+
+Note that, in this scheme, the last (k+1)-th subtotal is computed by time max_k+1 = max_k + 1, 
+where max_k is the time by which the last k-th subtotal become available. Since the last pair of 
+bags is distributed by time t=ceil(N/2), it follows that we can total N bags in, at worst, 
+ceil(N/2) + 2*ceil(log_2(N)) - 1 seconds. 
+
+*********************************
+In this case, it takes 143 seconds to total 256 bags using our scheme.
```

Comment:
I claimed it would take 133 seconds. This is because I drew out the cashiers and indexed them. Basically, every time a cashier gets a bag, he will spend an extra second summing it. We only need 128 cashiers since each cashier gets 2 bags. In second 2, cashier 1 receives his bags. Every second beyond that, we will have a cashier summed his bags. In seconds 4 and 5, cashier 1 and cashier 2 will communicate and sum their bags. From here, whenever a cashier sums the bags he received, he will sum with one of the two sets of people who just finished summing their bags. cashier ontinuing with the example, after second 5 (when cashier 4 finished summing his bags), cashier 1 will have the sum of his bags and cashier 2's bags, and sum that result with cashier 4's result. This will continue with the rest of the even indexed cashiers. This follows for odd indexed cashiers as well. By the end, we have cashier 128 sum his bags at the 129th second, and he communicates and sums that with two cashiers holding resulting sums. This process ends at 133 seconds.